Un-unsafe the `StableOrd` trait

eggyal opened the pull request:
Whilst incorrect implementations of this trait can cause miscompilation, they cannot cause memory unsafety in rustc.

[Discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Policy.20of.20.60unsafe.60.20within.20the.20compiler).

cc [MCP 533](https://github.com/rust-lang/compiler-team/issues/533), #105175, @michaelwoerister

r? @Nilstrieb

eggyal committed:
Un-unsafe the `StableOrd` trait

Whilst incorrect implementations of this trait can cause miscompilation,
they cannot cause memory unsafety in rustc.

michaelwoerister commented:
Yes, I agree that this shouldn't be `unsafe`. Implementing the trait has nothing to do with language invariants. @Nilstrieb was right all along 🙂

It would be nice though if we had some kind of `#[rustc::triple_check_every_impl_of_this_trait]` and `#[rustc::this_impl_has_been_triple_checked]` functionality.

eggyal commented:
Perhaps just add an associated `THIS_IMPLEMENTATION_HAS_BEEN_TRIPLE_CHECKED` constant to the trait?

eggyal commented:
However none of these approaches prevent the type later being modified to then have fields that break the invariant.

Perhaps a more robust approach would be to define it as an auto-trait that is unimplemented for types that are unstable across compilations?

Urgau commented:
> It would be nice though if we had some kind of `#[rustc::triple_check_every_impl_of_this_trait]` and `#[rustc::this_impl_has_been_triple_checked]` functionality.

This feels very similar to the proposed the [`clippy::dangerous`](https://github.com/rust-lang/rust-clippy/pull/11600) attribute. I wish it were already merged and usable.

michaelwoerister commented:
`[rustc::this_impl_has_been_triple_checked_for_transitive_hash_of_all_reachable_code(0x124159831948129481821948)]` 😉

michaelwoerister commented:
> Perhaps just add an associated THIS_IMPLEMENTATION_HAS_BEEN_TRIPLE_CHECKED constant to the trait?

That seems as effective as marking as unsafe, at least 👍

michaelwoerister commented:
> Perhaps a more robust approach would be to define it as an auto-trait that is unimplemented for types that are unstable across compilations?

If we can reliably define that set of types, that would be the best solution.

michaelwoerister commented:
However, we also have to take the implementation of `Ord` into account which can do whatever it wants.

Nilstrieb commented:
i feel like this is overcomplicated it. keeping it as-is as a normal trait and maybe having a "I promise it's fine" const should be more than enough.

eggyal committed:
Ensure careful consideration is given by impls

Added an associated `const THIS_IMPLEMENTATION_HAS_BEEN_TRIPLE_CHECKED`
to the `StableOrd` trait to ensure that implementors carefully consider
whether the trait's contract is upheld, as incorrect implementations can
cause miscompilations.

eggyal commented:
@rustbot ready

michaelwoerister commented:
Feel free to r=me, unless @Nilstrieb has any further comments.

Nilstrieb commented:
i don't!
@bors r=michaelwoerister

bors commented:
:pushpin: Commit 0e73e7095ae7aa7ead69c4ba7ba002560ef118fa has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister 0e73e7095ae7aa7ead69c4ba7ba002560ef118fa -->
<!-- homu: {"type":"Approved","sha":"0e73e7095ae7aa7ead69c4ba7ba002560ef118fa","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

michaelwoerister commented:
Thanks for cleaning this up, @eggyal 🙂

bors commented:
:hourglass: Testing commit 0e73e7095ae7aa7ead69c4ba7ba002560ef118fa with merge c290e9de32e8ba6a673ef125fde40eadd395d170...
<!-- homu: {"type":"BuildStarted","head_sha":"0e73e7095ae7aa7ead69c4ba7ba002560ef118fa","merge_sha":"c290e9de32e8ba6a673ef125fde40eadd395d170"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9665666246/job/26669099283)
Approved by: michaelwoerister
Pushing c290e9de32e8ba6a673ef125fde40eadd395d170 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9665666246/job/26669099283"},"merge_sha":"c290e9de32e8ba6a673ef125fde40eadd395d170"} -->

rust-timer commented:
Finished benchmarking commit (c290e9de32e8ba6a673ef125fde40eadd395d170): [comparison URL](https://perf.rust-lang.org/compare.html?start=d929a42a664c026167800801b26d734db925314f&end=c290e9de32e8ba6a673ef125fde40eadd395d170&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d929a42a664c026167800801b26d734db925314f&end=c290e9de32e8ba6a673ef125fde40eadd395d170&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [0.3%, 5.7%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d929a42a664c026167800801b26d734db925314f&end=c290e9de32e8ba6a673ef125fde40eadd395d170&stat=max-rss)
<details>
<summary>Results (primary 2.4%, secondary 6.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4% | [2.4%, 2.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | 6.0% | [4.5%, 8.2%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.4% | [2.4%, 2.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d929a42a664c026167800801b26d734db925314f&end=c290e9de32e8ba6a673ef125fde40eadd395d170&stat=cycles:u)
<details>
<summary>Results (secondary 3.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8% | [3.8%, 3.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d929a42a664c026167800801b26d734db925314f&end=c290e9de32e8ba6a673ef125fde40eadd395d170&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 692.703s -> 694.374s (0.24%)
**Artifact size**: 326.77 MiB -> 324.66 MiB (-0.64%)
